pandas/tests/test_strings.py

```diff
@@ -420,6 +420,16 @@ def test_str_cat_categorical(self, box, dtype_caller, dtype_target, sep):
             result = s.str.cat(t, sep=sep)
             assert_series_or_index_equal(result, expected)
 
+    @pytest.mark.parametrize('box', [Series, Index, np.array, list])
```

Comment:
do these cover the different conditions for raising?

Reply:
IMO yes. It would be possible to add more dtypes, but I don't consider this a great improvement.

Happy to add if you don't mind combinatorially more tests.

Reply:
so if i modify the condition, we have tests that fail?

Reply:
I don't understand what you mean here. Adding different containers (as long as they are legal inputs to `.str.cat`) is possible, but I've already added the most import ones.

It would be possible add a further parametrization over the dtype and not just test that integer Series fail (as currently), but also floats, complex, datetime etc. This would work without failure.

You'd have to be  more concrete with "modify the condition" for me to answer more clearly. This test serves as a relatively minimal example from #22722. If you want more dtypes, I'm happy to add them. Anything else you'll need to explain to me in more detail.

Reply:
not raising the message on an invalid case would just result in the old message be raised. rasing on valid cases is a regression.

would it not be simpler to catch the message and reraise?

Reply:
> would it not be simpler to catch the message and reraise?

In principle, yes. But:
* there are 3 separate calls to `cat_core` that would need to be wrapped
* performance: I don't want to do all the gymnastics of aligning, `ensure_object`, calculating and unioning na_masks and cat_core itself if we already (could) know beforehand that things will fail.

> not raising the message on an invalid case would just result in the old message be raised. rasing on valid cases is a regression.

It is possible to circumvent the check I'm adding. `pd.Series(['a', 1, 'b'])` will fail with the old error, but we wouldn't be able to tell from the dtype beforehand. Raising on valid cases cannot happen I believe - all valid concatenations must be either strings or missing values (that's why the all-NA float case is in there, which can happen for example due to mis-alignment; as opposed to e.g. complex, which could have all-NA but shouldn't [be expected to] work with strings)

Reply:
@h-vetinari thanks for the explanation. i'm not convinced that for the sake of a better error message we should be adding this complexity, especially if the old message is still raised in some cases.

@jreback

Reply:
> i'm not convinced that for the sake of a better error message we should be adding this complexity, especially if the old message is still raised in some cases.

I think looking at #22722, this is a clear win - it covers 99% (guesstimate) of real-world cases where this error happens. If someone actually has mixed strings / integers in their Series, it just won't fail as gracefully. But that doesn't imply (to me) that we shouldn't aim to do better in the 99%.

Reply:
Side note: if your only worry is an inconsistent error message, I could have the 99%-fail-early-check *and* catch-and-reraise afterwards (with the same error message).

Reply:
best to wait for feedback from other maintainers before you make any changes on my account.

Reply:
Too late. ;-)